File: macos/EMWaver/EMWaver/DeviceRegistryService.swift
import Combine
import Foundation

/// When a genuine device connects (DeviceID+Proof verified locally), inform backend so it can be attached
/// to the signed-in account (or prompt for login if user is anonymous).
@MainActor
final class DeviceRegistryService: ObservableObject {
    private var cancellables: Set<AnyCancellable> = []
    private var lastSeenKey: String = ""
    private weak var accountDevices: AccountDevicesService?

    func start(auth: AuthenticationManager, device: MacUSBManager, accountDevices: AccountDevicesService) {
        self.accountDevices = accountDevices
        // Re-check whenever auth or secure device identity changes.
        Publishers.CombineLatest(auth.$session, device.$secureDeviceIdB64)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] session, deviceIdB64 in
                guard let self else { return }
                Task { await self.sync(auth: auth, device: device, session: session, deviceIdB64: deviceIdB64) }
            }
            .store(in: &cancellables)
    }

    private func sync(auth: AuthenticationManager, device: MacUSBManager, session: AuthSession?, deviceIdB64: String?) async {
        guard device.isConnected, device.isSecureConnected else { return }
        guard let deviceIdB64, !deviceIdB64.isEmpty else { return }
        guard let proofB64 = device.secureDeviceProofB64, !proofB64.isEmpty else { return }
        guard let baseURL = BackendUrl.resolve() else {
            device.deviceAttachStatusText = "Missing backend URL"
            return
        }

        let token = (session?.idToken ?? "")
        let key = "\(deviceIdB64):\(token.isEmpty ? "anon" : "auth")"
        if key == lastSeenKey { return }
        lastSeenKey = key

        device.deviceAttachStatusText = "Checking device…"

        var req = URLRequest(url: baseURL.appendingPathComponent("/v1/devices/seen"))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if !token.isEmpty {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let payload: [String: Any] = ["device_id_b64": deviceIdB64, "proof_b64": proofB64]
        req.httpBody = try? JSONSerialization.data(withJSONObject: payload)

        do {
            let (data, resp) = try await URLSession.shared.data(for: req)
            guard let http = resp as? HTTPURLResponse else {
                device.deviceAttachStatusText = "Device check failed"
                return
            }

            let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]

            if http.statusCode != 200 {
                let raw = String(data: data, encoding: .utf8) ?? ""
                device.deviceAttachStatusText = raw.isEmpty ? "Device check failed (HTTP \(http.statusCode))" : raw
                return
            }

            let needsLogin = (obj?["needs_login"] as? Bool) ?? false
            let attached = (obj?["attached"] as? Bool) ?? false
            let claimed = (obj?["claimed"] as? Bool) ?? false

            if needsLogin && token.isEmpty {
                let suppress = UserDefaults.standard.bool(forKey: "emwaver.deviceAttachPrompt.suppress")
                if !suppress {
                    device.needsLoginToSaveDevice = true
                }
                device.deviceAttachStatusText = "Sign in to save device"
                return
            }

            device.needsLoginToSaveDevice = false
            if attached {
                device.deviceAttachStatusText = "Device saved to account"
            } else if claimed {
                device.deviceAttachStatusText = "Device verified"
            } else {
                device.deviceAttachStatusText = "Device verified"
            }

            if let hardwareUid = device.hardwareUidHex,
               !hardwareUid.isEmpty,
               let boardType = device.connectedBoardType,
               !boardType.isEmpty {
                accountDevices?.storeClaimedDevice(
                    deviceIdB64: deviceIdB64,
                    boardType: boardType,
                    hardwareUid: hardwareUid
                )
            }
        } catch {
            device.deviceAttachStatusText = error.localizedDescription
        }
    }
}

File: windows/EMWaver/Services/DeviceRegistryService.cs
using EMWaver.Services.Cloud;
using System;
using System.ComponentModel;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace EMWaver.Services;

internal sealed class DeviceRegistryService
{
    private readonly HttpClient _http;
    private readonly Func<CloudAuthManager> _auth;
    private readonly Func<AccountDevicesService> _accountDevices;
    private readonly SemaphoreSlim _syncLock = new(1, 1);

    private WindowsDeviceManager? _device;
    private string _lastSeenKey = "";
    private bool _started;
    private CloudAuthManager? _subscribedAuth;

    internal DeviceRegistryService(HttpClient http, Func<CloudAuthManager> auth, Func<AccountDevicesService> accountDevices)
    {
        _http = http;
        _auth = auth;
        _accountDevices = accountDevices;
    }

    internal void Start(WindowsDeviceManager device)
    {
        if (_started) return;
        _started = true;
        _device = device;

        device.PropertyChanged += OnDevicePropertyChanged;
        EnsureAuthSubscription();
        QueueSync();
    }

    private void OnDevicePropertyChanged(object? sender, PropertyChangedEventArgs e)
    {
        if (e.PropertyName == nameof(WindowsDeviceManager.IsConnected) ||
            e.PropertyName == nameof(WindowsDeviceManager.IsSecureConnected) ||
            e.PropertyName == nameof(WindowsDeviceManager.SecureDeviceIdB64) ||
            e.PropertyName == nameof(WindowsDeviceManager.SecureDeviceProofB64) ||
            e.PropertyName == nameof(WindowsDeviceManager.HardwareUidHex) ||
            e.PropertyName == nameof(WindowsDeviceManager.ConnectedBoardType))
        {
            QueueSync();
        }
    }

    private void OnAuthChanged()
    {
        QueueSync();
    }

    private void QueueSync()
    {
        EnsureAuthSubscription();
        _ = SyncAsync();
    }

    private void EnsureAuthSubscription()
    {
        var auth = _auth();
        if (ReferenceEquals(_subscribedAuth, auth))
        {
            return;
        }

        if (_subscribedAuth != null)
        {
            _subscribedAuth.Changed -= OnAuthChanged;
        }

        _subscribedAuth = auth;
        _subscribedAuth.Changed += OnAuthChanged;
    }

    private async Task SyncAsync()
    {
        var device = _device;
        if (device == null) return;
        if (!device.IsConnected || !device.IsSecureConnected) return;
        if (string.IsNullOrWhiteSpace(device.SecureDeviceIdB64) || string.IsNullOrWhiteSpace(device.SecureDeviceProofB64)) return;

        await _syncLock.WaitAsync();
        try
        {
            var token = await _auth().GetValidIdTokenAsync(CancellationToken.None, interactiveSignIn: false) ?? "";
            var key = $"{device.SecureDeviceIdB64}:{(string.IsNullOrWhiteSpace(token) ? "anon" : "auth")}";
            if (string.Equals(_lastSeenKey, key, StringComparison.Ordinal))
            {
                return;
            }

            _lastSeenKey = key;
            device.DeviceAttachStatusText = "Checking device...";

            var url = $"{BackendUrl.Resolve().TrimEnd('/')}/v1/devices/seen";
            using var req = new HttpRequestMessage(HttpMethod.Post, url);
            req.Headers.Accept.ParseAdd("application/json");
            if (!string.IsNullOrWhiteSpace(token))
            {
                req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
            }

            var payload = JsonSerializer.Serialize(new
            {
                device_id_b64 = device.SecureDeviceIdB64,
                proof_b64 = device.SecureDeviceProofB64,
            });
            req.Content = new StringContent(payload, Encoding.UTF8, "application/json");

            using var res = await _http.SendAsync(req);
            var body = await res.Content.ReadAsStringAsync();
            if (!res.IsSuccessStatusCode)
            {
                device.DeviceAttachStatusText = string.IsNullOrWhiteSpace(body)
                    ? $"Device check failed (HTTP {(int)res.StatusCode})"
                    : body;
                return;
            }

            using var doc = JsonDocument.Parse(string.IsNullOrWhiteSpace(body) ? "{}" : body);
            var root = doc.RootElement;

            var needsLogin = root.TryGetProperty("needs_login", out var needsLoginEl) && needsLoginEl.GetBoolean();
            var attached = root.TryGetProperty("attached", out var attachedEl) && attachedEl.GetBoolean();
            var claimed = root.TryGetProperty("claimed", out var claimedEl) && claimedEl.GetBoolean();

            if (needsLogin && string.IsNullOrWhiteSpace(token))
            {
                device.NeedsLoginToSaveDevice = true;
                device.DeviceAttachStatusText = "Sign in to save device";
                return;
            }

            device.NeedsLoginToSaveDevice = false;
            device.DeviceAttachStatusText = attached
                ? "Device saved to account"
                : (claimed ? "Device verified" : "Device verified");

            if (!string.IsNullOrWhiteSpace(device.HardwareUidHex) &&
                !string.IsNullOrWhiteSpace(device.ConnectedBoardType) &&
                !string.IsNullOrWhiteSpace(device.SecureDeviceIdB64))
            {
                _accountDevices().StoreClaimedDevice(
                    device.SecureDeviceIdB64!,
                    device.ConnectedBoardType!,
                    device.HardwareUidHex!);
            }
        }
        catch (Exception ex)
        {
            device.DeviceAttachStatusText = ex.Message;
        }
        finally
        {
            _syncLock.Release();
        }
    }
}
